### DISCUSSION TITLE
FedProx Implementation

### DISCUSSION THREAD

**Discussion by logjay** [MLOPTPSU/FedTorch#18] [Q&A] [answered]
https://github.com/MLOPTPSU/FedTorch/blob/ab8068dbc96804a5c1a8b898fd115175cfebfe75/fedtorch/comms/trainings/federated/centered/afl.py#L130

This line contributes to the loss minimizer, but loss.backward() is called before this. Don't we miss out on the effect of FedProx here from the backpropagation?

**Comment by mmkamani7** [comment, marked as answer]
The initial backward operation computes the gradient for each parameter. Subsequently, the mentioned line incorporates gradients from FedProx regularization and the loss value into this calculation. By employing `grad.data`, we're updating the gradient previously derived from the .backward() call, without disrupting its original computation.

**Reply by logjay** [reply]
So you're manually adding to the loss magnitude with line 130, while manually adjusting the gradient (using the derivative of this line) with line 131?

So it looks like:

loss = criterion()
- the loss is calculated

loss.backward()
- the backpropagation calculates the gradients from error

[loss, grad fedprox summation]
- the loss magnitude is increased
- the gradient is adjusted based on fedprox

optimizer.step()
- the weights are adjusted using the gradient and loss


Is this correct?

**Reply by mmkamani7** [reply]
Yes that is correct.
